stdlib/public/core/SmallString.swift

```diff
@@ -231,14 +216,28 @@ extension _SmallString {
       return try f(UnsafeMutableBufferPointer(
         start: ptr, count: _SmallString.capacity))
     }
-
     _internalInvariant(len <= _SmallString.capacity)
-    discriminator = .small(withCount: len, isASCII: self.computeIsASCII())
+
+    let (leading, trailing) = self.zeroTerminatedRawCodeUnits
+    self = _SmallString(leading: leading, trailing: trailing, count: len)
   }
 }
 
 // Creation
 extension _SmallString {
+  @inlinable @inline(__always)
+  internal init(leading: UInt64, trailing: UInt64, count: Int) {
+    _internalInvariant(count <= _SmallString.capacity)
+
+    let isASCII = (leading | trailing) & 0x8080_8080_8080_8080 == 0
```

Comment:
Ooh, nice. (I know it's not new; I just hadn't seen it before.)

Reply:
Actually, given how fast this is, I wonder if the dedicated `isASCII` bit is even needed, or if it's worth saving for something.

Reply:
It's not that fast; we have to mask off the discriminator too. While ARM can encode the constant as an immediate, but it's still more code. Specifically for small strings, ASCII-ness is by far the most common and important characteristic to know.

The other contender for this bit would be remembering both isASCII and doesn't have CR+LF, i.e. single-scalar-grapheme. But, I think it's better to check that on Character access.

```swift
  public func _forCodeGen_1() { if isASCII { foo() } else { bar() } }
  public func _forCodeGen_2() { if computeIsASCII() { foo() } else { bar() } }
```

```
_$ss12_SmallStringV13_forCodeGen_1yyF:
  tbnz       x1, 0x3e, loc_15c474

  b          _$ss12_SmallStringV3baryyFTf4d_n

loc_15c474:
  b          _$ss12_SmallStringV3fooyyFTf4d_n



_$ss12_SmallStringV13_forCodeGen_2yyF:
  orr        x8, xzr, #0x8080808080808080
  movk       x8, #0x80, lsl #48
  and        x8, x1, x8
  and        x9, x0, #0x8080808080808080
  orr        x8, x8, x9
  cbz        x8, loc_15c494

  b          _$ss12_SmallStringV3baryyFTf4d_n
loc_15c494:
  b          _$ss12_SmallStringV3fooyyFTf4d_n
```

Considering that this check should probably be at the top of every String API inlined into user code, it's probably worth it.

Reply:
In theory that could be

```
mov x8, x0, lsl #8
orr x8, x1, x8
and x8, x8, #0x8080808080808080
cbz x8, loc
```

but I take your point.